From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RemainingToast <59129680+RemainingToast@users.noreply.github.com>
Date: Wed, 15 Dec 2021 19:39:21 +1000
Subject: [PATCH] Map Decoration Limit

The root cause is the map icon not being removed from the decorations hash map after the item frame is removed.

Don't count some map decorations

TODO: move modifications to other patches out of this patch.

diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index f377b0beab436cfdc4694895cf8ed19644ee8c6e..d614d6e28963b8aff2e4088734115a5492699a7a 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -163,9 +163,9 @@ public class DionysusConfig {
         return config.getStringList(key);
     }
 
-    public static long maxLightTimePerTick;
+    public static int maxLightTimePerTick = 10;
     private static void maxLightTimePerTickMs() {
-        maxLightTimePerTick = getInt("light-updates-max-ms-per-tick", 30) * 1000000L;
+        maxLightTimePerTick = getInt("light-updates-max-ms-per-tick", maxLightTimePerTick) * 1000000;
     }
 
     public static boolean customNameAiLimitBypass = true;
@@ -231,12 +231,17 @@ public class DionysusConfig {
 
     public static boolean alternativeBlockPlacement = true;
     private static void alternativeBlockPlacement() {
-        alternativeBedMechanics = getBoolean("tickless-block-placement", alternativeBlockPlacement);
+        alternativeBlockPlacement = getBoolean("tickless-block-placement", alternativeBlockPlacement);
         if (alternativeBlockPlacement) {
             Bukkit.getLogger().info("Tickless Block Placement Mechanics activated. Block placement based on time instead of ticks.");
         }
     }
 
+    public static int mapDecorationsLimit = 10;
+    private static void mapDecorationsLimit() {
+        mapDecorationsLimit = getInt("map-decoration-limit", mapDecorationsLimit);
+    }
+
     public static boolean chunkOverloadDupe = false;
     public static boolean chunkOverloadDupeLog = false;
     private static void chunkOverloadDupe() {
@@ -245,4 +250,13 @@ public class DionysusConfig {
         chunkOverloadDupeLog = getBoolean("chunk-overload-dupe.log", chunkOverloadDupeLog);
         if (chunkOverloadDupeLog) Bukkit.getLogger().info("Chunk overload dupes will be logged.");
     }
+
+    public static int tileEntityDataLimit = 750;
+    public static int tileEntitySplitLimit = 15000;
+    private static void tileEntityLimit() {
+        tileEntityDataLimit = getInt("tile-entity.data-limit", tileEntityDataLimit);
+        tileEntitySplitLimit = getInt("tile-entity.split-limit", tileEntitySplitLimit);
+    }
+
+
 }
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index a4eddca67a7c401da8351c2e994d14119e96d692..56b08a0a417c2d47728b62d78ec0db3906c87b47 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -48,13 +48,13 @@ public abstract class Entity implements ICommandListener, KeyedObject { // Paper
 
     // CraftBukkit start
     private static final int CURRENT_LEVEL = 2;
-    // Paper start
+    // Dionysus - Paper start
     public static Random SHARED_RANDOM = new XoShiRo256PlusPlusRandom() {
         private boolean locked = false;
         @Override
         public synchronized void setSeed(long seed) {
             if (locked) {
-                LogManager.getLogger().error("Ignoring setSeed on Entity.SHARED_RANDOM", new Throwable());
+                LogManager.getLogger().debug("Ignoring setSeed on Entity.SHARED_RANDOM", new Throwable());
             } else {
                 super.setSeed(seed);
                 locked = true;
@@ -66,8 +66,10 @@ public abstract class Entity implements ICommandListener, KeyedObject { // Paper
             return this.nextDoubleFast();
         }
     };
+    // Dionysus end
     List<Entity> entitySlice = null;
     // Paper end
+
     static boolean isLevelAtLeast(NBTTagCompound tag, int level) {
         return tag.hasKey("Bukkit.updateLevel") && tag.getInt("Bukkit.updateLevel") >= level;
     }
diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
index 42d7dff874ecac0d5760dc4f9b6c13e6f078f408..30a2097975868ff80232cd0c32d7d55b0c374aa1 100644
--- a/src/main/java/net/minecraft/server/LoginListener.java
+++ b/src/main/java/net/minecraft/server/LoginListener.java
@@ -188,9 +188,13 @@ public class LoginListener implements PacketLoginInListener, ITickable {
 
     public String c() {
         String address = this.networkManager.getSocketAddress().toString();
+
+        // Dionysus start
         if (DionysusConfig.redactPlayerData) {
             address = "/0.0.0.0:0";
         }
+        // Dionysus end
+
         return this.i != null ? this.i + " (" + address + ")" : address;
     }
 
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index df8940ff1c317ca321e0a03d06eafffd4b561c6c..4c6de8a38d7d90ad9cfe0192c2d08d07f349e52a 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -761,8 +761,8 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
     protected void C() throws ExceptionWorldConflict { // CraftBukkit - added throws
         co.aikar.timings.TimingsManager.FULL_SERVER_TICK.startTiming(); // Paper
         this.slackActivityAccountant.tickStarted(); // Spigot
-        long i = System.nanoTime(); long startTime = i; // Paper
-        tickStartTime = startTime; // Dionysus
+        long i = System.nanoTime();
+        tickStartTime = i; // Dionysus
 
         ++this.ticks;
         if (this.T) {
@@ -831,7 +831,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
         this.methodProfiler.b();
 
         org.spigotmc.WatchdogThread.tick(); // Spigot
-        DionysusLightingQueue.processQueue(startTime); // Paper
+        DionysusLightingQueue.processQueue(i); // Paper
         this.slackActivityAccountant.tickEnded(tickNanos); // Spigot
         co.aikar.timings.TimingsManager.FULL_SERVER_TICK.stopTiming(); // Paper
     }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
index 108d4c500977ed34431b57fbf35feb36a43660ba..54042bee18a7369322e7c1122b5d6667eaf05378 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
@@ -1,6 +1,7 @@
 package net.minecraft.server;
 
 import com.google.common.collect.Lists;
+import dev.pomf.dionysus.DionysusConfig;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
 import java.io.IOException;
@@ -30,10 +31,6 @@ public class PacketPlayOutMapChunk implements Packet<PacketListenerPlayOut> {
 
     // Paper start
     private final java.util.List<Packet> extraPackets = new java.util.ArrayList<>();
-    private static final int TE_LIMIT = Integer.getInteger("dionysus.te-limit", 750); // Dionysus - Tuinity - handle oversized chunk data packets more robustly
-    private static final int TE_SPLIT_LIMIT = Math.max(4096 + 1, Integer.getInteger("dionysus.te-split-limit", 15_000)); // Dionysus - Tuinity- handle oversized chunk data packets more robustly
-    private boolean mustSplit = false; // Dionysus - Tuinity - handle oversized chunk data packets more robustly
-
     public java.util.List<Packet> getExtraPackets() {
         return extraPackets;
     }
@@ -57,7 +54,11 @@ public class PacketPlayOutMapChunk implements Packet<PacketListenerPlayOut> {
         this.c = this.writeChunk(new PacketDataSerializer(this.g()), chunk, flag, i, packetPlayOutMapChunkInfo); // Paper - Anti-Xray - Add chunk packet info
         this.e = Lists.newArrayList();
         Iterator iterator = chunk.getTileEntities().entrySet().iterator();
-        int totalTileEntities = 0; // Dionysus
+
+        // Dionysus start
+        int totalTileEntities = 0;
+        boolean mustSplit = false;
+        // Dionysus end
 
         while (iterator.hasNext()) {
             Entry entry = (Entry) iterator.next();
@@ -68,13 +69,13 @@ public class PacketPlayOutMapChunk implements Packet<PacketListenerPlayOut> {
             if (this.e() || (i & 1 << j) != 0) {
                 // Dionysus Start - handle tile entities better
                 ++totalTileEntities;
-                if (totalTileEntities > TE_SPLIT_LIMIT) {
-                    this.mustSplit = true;
+                if (totalTileEntities > DionysusConfig.tileEntitySplitLimit) {
+                    mustSplit = true;
                     this.getTileDataEntityData().clear();
                     this.extraPackets.clear();
                     break;
                 }
-                if (totalTileEntities > TE_LIMIT) {
+                if (totalTileEntities > DionysusConfig.tileEntityDataLimit) {
                     PacketPlayOutTileEntityData updatePacket = tileentity.getUpdatePacket();
                     if (updatePacket != null) {
                         extraPackets.add(updatePacket);
@@ -90,7 +91,7 @@ public class PacketPlayOutMapChunk implements Packet<PacketListenerPlayOut> {
 
         chunk.world.chunkPacketBlockController.modifyBlocks(this, packetPlayOutMapChunkInfo); // Paper - Anti-Xray - Modify blocks
         // Dionysus - Tuinity start - improve oversized chunk data packet handling
-        if (this.mustSplit) {
+        if (mustSplit) {
             int chunkSectionBitSetCopy = chunkSectionBitSet;
             for (int a = 0, len = Integer.bitCount(chunkSectionBitSet); a < len; ++a) {
                 // IntegerUtil.getTrailingBit (-n & n)
diff --git a/src/main/java/net/minecraft/server/WorldMap.java b/src/main/java/net/minecraft/server/WorldMap.java
index 59173605ee8406b021576329657a8f1477fe10ad..9d34b65f21e9b8a240e290433cefba14071035b8 100644
--- a/src/main/java/net/minecraft/server/WorldMap.java
+++ b/src/main/java/net/minecraft/server/WorldMap.java
@@ -10,12 +10,16 @@ import javax.annotation.Nullable;
 
 // CraftBukkit start
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
 
+import dev.pomf.dionysus.DionysusConfig;
+import org.bukkit.Bukkit;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.map.CraftMapView;
 // CraftBukkit end
 
+
 public class WorldMap extends PersistentBase {
 
     public int centerX;
@@ -282,9 +286,40 @@ public class WorldMap extends PersistentBase {
             }
         }
 
+        // Dionysus start - limit decorations
+        if (countMapIcon(this.decorations) >= DionysusConfig.mapDecorationsLimit) {
+            return;
+        }
+
+        // Dionysus end
         this.decorations.put(s, new MapIcon(mapicon_type, b0, b1, b2));
     }
 
+    // Dionysus start
+    // Not all map decorations need to be prevented
+    // See https://minecraft.fandom.com/wiki/Map#Map_icons
+    private int countMapIcon(Map<UUID, MapIcon> icons) {
+        AtomicInteger count = new AtomicInteger();
+
+        for (MapIcon icon : icons.values()) {
+            switch (MapIcon.Type.a(icon.getType())) {
+                case PLAYER_OFF_MAP:
+                case PLAYER_OFF_LIMITS:
+                case PLAYER:
+                case MANSION:
+                case MONUMENT: {
+                    continue;
+                }
+                default: {
+                    count.getAndIncrement();
+                }
+            }
+        }
+
+        return count.get();
+    }
+    // Dionysus end
+
     @Nullable
     public Packet<?> a(ItemStack itemstack, World world, EntityHuman entityhuman) {
         WorldMap.WorldMapHumanTracker worldmap_worldmaphumantracker = (WorldMap.WorldMapHumanTracker) this.k.get(entityhuman);
